### PR TITLE
macOS clang ASAN support

### PR DESCRIPTION
**Pull request by andyp1per**
Adds support for building with the macOS clang Address Sanitizer

### CODE REVIEW

**Comment by peterbarker** [comment]
What makes this MacOSX-specific?

**Comment by andyp1per** [comment]
> What makes this MacOSX-specific?

You are right - it is clang specific

File: Tools/ardupilotwaf/boards.py [hunk 1]
```diff
@@ -368,6 +368,11 @@ def configure_env(self, cfg, env):
             env.CXXFLAGS += [
                 '-O3',
             ]
+        elif 'clang++' in cfg.env.COMPILER_CXX:
```
Comment (peterbarker) [inline Tools/ardupilotwaf/boards.py]: Let's say "no" to double-negatives :-)  Please remove the `not` in the `if not`

**Comment by andyp1per** [comment]
> 
> 
> I like the idea.
> 
> Any extra packages or whatnot need to be installed for asan?

You have to have the llvm symbolizer in your path (llvm-symbolizer) which is not installed by default on macos. Easiest way to get it is to ``brew install llvm``

I was thinking of controlling this via a waf ```--asan`` option. Thoughts?

**Comment by peterbarker** [comment]
On Sun, 12 Jan 2020, Andy Piper wrote:

> I was thinking of controlling this via a waf ```--asan`` option. Thoughts?

Yep, something like that will be required.

File: Tools/ardupilotwaf/boards.py (hunk 1 above)
Comment (andyp1per) [reply Tools/ardupilotwaf/boards.py]: Fixed

**Comment by peterbarker** [comment]
```
pbarker@bluebottle:~/rc/ardupilot((HEAD detached at andyp1per/pr-asan-support))$ CXX=clang++ ./waf configure --asan
pbarker@bluebottle:~/rc/ardupilot((HEAD detached at andyp1per/pr-asan-support))$ ./waf build --target=bin/arducopter
```

Yields:

```
In file included from ../../libraries/SITL/SIM_EFI_MegaSquirt.cpp:21:
In file included from ../../libraries/AP_HAL/utility/sparse-endian.h:23:
../../libraries/AP_Common/missing/byteswap.h:9:24: fatal error: expected unqualified-id
static inline uint16_t __bswap_16(uint16_t u)
                       ^
/usr/include/x86_64-linux-gnu/bits/byteswap-16.h:26:7: note: expanded from macro '__bswap_16'
     (__extension__                                                           \
      ^
1 error generated.

Waf: Leaving directory `/home/pbarker/rc/ardupilot/build/sitl'
Build failed
 -> task in 'objs/SITL' failed (exit status 1): 
	{task 140368177583840: cxx SIM_EFI_MegaSquirt.cpp -> SIM_EFI_MegaSquirt.cpp.0.o}
 (run with -v to display more information)
```

**Comment by andyp1per** [comment]
I'm guessing this is linux specific as it works on my machine. What version of clang are you using?

**Comment by peterbarker** [comment]
On Tue, 14 Jan 2020, Andy Piper wrote:

> I'm guessing this is linux specific as it works on my machine. What version of clang are you using?

pbarker@bluebottle:~$ clang++ --version
clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
pbarker@bluebottle:~$

**Comment by andyp1per** [comment]
@peterbarker I think you are mixing compilers. I tried in bionic64 with this command line and all was good:
```
CXX=clang++ CC=clang ./waf configure --board sitl
```

**Comment by peterbarker** [comment]
> @peterbarker I think you are mixing compilers. I tried in bionic64 with this command line and all was good:

Quite right.  Never been a problem before, and that fixes that problem.

**Comment by peterbarker** [comment]
OK, so I had a go at playing with this today.

Sadly, things did not go well.

The main loop starts, but we never get sensor samples, because the time fields for next-sample on our simulated gyros and accels are some very large number.

Now we override the "new" function to zero the allocated memory.

My guess is that asan is stopping that override from working - so this doesn't appear to be a goer.

@andyp1per am I missing something about how asan is supposed to be run?

My test branch is here: https://github.com/peterbarker/ardupilot/tree/pr/asan-support

**Comment by andyp1per** [comment]
@peterbarker pretty sure this is working on my mac, but on my mac I am using clang version 9.0. Any chance you can try this?

**Comment by peterbarker** [comment]
Sorry, can't talk for clang-9, but clang-8 suffers from the same problem - uninitialised memory.

Some sort of error message from waf that you MUST have `--debug` enabled for `--asan` to compile is warranted, I think.  That was a bit of a time sink for me :-P  That was actually my original problem with byteswap perhaps - not that I was mixing compilers but that things weren't compiling and it wasn't getting the byteswap.h include correct.

**Comment by andyp1per** [comment]
> Sorry, can't talk for clang-9, but clang-8 suffers from the same problem - uninitialised memory.
> 
> Some sort of error message from waf that you MUST have `--debug` enabled for `--asan` to compile is warranted, I think. That was a bit of a time sink for me :-P That was actually my original problem with byteswap perhaps - not that I was mixing compilers but that things weren't compiling and it wasn't getting the byteswap.h include correct.

Hmmn, how are you testing this? I can't get things to break at all - I'm running ```./Tools/autotest/autotest.py fly.ArduPlane.FRSkyPassThrough``` as a test. Note though that I just rebased and saw the FRSkyD hang again locally, so maybe the stack issue is not fully resolved?

**Comment by peterbarker** [comment]
Just tried it on clang9 - same issue.

Could you see this for yourself by running it on Ubuntu?

**Comment by andyp1per** [comment]
@peterbarker I'm not having much success in vagrant - which is all I have - but not sure this is the same problem as you:
```
Log Directory: test.ArduPlane/logs/2020-01-16/flight2
Telemetry log: test.ArduPlane/logs/2020-01-16/flight2/flight.tlog
AUTOTEST: LOGFILE test.ArduPlane/logs/2020-01-16/flight2/flight.tlog
AUTOTEST: buildlog=/tmp/buildlogs/ArduPlane-test.tlog
AUTOTEST: OSError [18]: Invalid cross-device link
AUTOTEST: WARN: Failed to create symlink: test.ArduPlane/logs/2020-01-16/flight2/flight.tlog => /tmp/buildlogs/ArduPlane-test.tlog, will copy tlog manually to target location
AUTOTEST: Waiting for Parameters
MAV> Waiting for heartbeat from tcp:127.0.0.1:5760
link 1 down
```

**Comment by peterbarker** [comment]
On Thu, 16 Jan 2020, Andy Piper wrote:

> @peterbarker I'm not having much success in vagrant - which is all I have -
> but not sure this is the same problem as you:

> AUTOTEST: Waiting for Parameters
> MAV> Waiting for heartbeat from tcp:127.0.0.1:5760
> link 1 down

Nope, that's probably it.  It's a bit tough to diagnose, but if you put a 
break in AP_InertialSensor wait_for_sample you can trace through and see 
that the next time  a sample is due from the simulated sensors is a *long* 
way into the future.  Actually, a breakpoint just after we "new" the fake 
sensors works well.

**Comment by andyp1per** [comment]
@peterbarker https://github.com/google/sanitizers/wiki/AddressSanitizer
I'm guessing this is what you are seeing?:



Q: What do 0xbebebebebebebebe and 0xbebebebe mean?

A: ASan, by default, writes 0xbe to newly allocated memory (see malloc_fill_byte). 0xbebebebebebebebe is (possibly) a 64-bit value that was allocated but not initialized. Similarly for 0xbebebebe.

**Comment by andyp1per** [comment]
When we build with clang do we use clang to link as well?

**Comment by andyp1per** [comment]
> Nope, that's probably it. It's a bit tough to diagnose, but if you put a break in AP_InertialSensor wait_for_sample you can trace through and see that the next time a sample is due from the simulated sensors is a *long* way into the future. Actually, a breakpoint just after we "new" the fake sensors works well.

@peterbarker actually it turns out this is nothing to do with ASAN at all - clang itself doesn't work on these tests, but gcc does which makes me think there is still something wrong in the threading or FRSky code. If I build with clang they fail, build with gcc and they pass.

**Comment by peterbarker** [comment]
@andyp1per Just ran all three Frsky tests under clang here without error.  Which version of clang++?  Don't suppose it's a Vagrant VM by any chance? :-)

**Comment by andyp1per** [comment]
@peterbarker I was using clang 6 I think. yes it was a vagrant vm 😄

**Comment by peterbarker** [comment]
On Fri, 24 Jan 2020, Andy Piper wrote:

> @peterbarker I was using clang 6 I think. yes it was a vagrant vm 😄

pbarker@bluebottle:~/rc/ardupilot(pr/onboard-cal-improvements)$ clang++ 
--version
clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin


So that's a bit vexing.  Don't suppose you remember which particular VM?

**Comment by andyp1per** [comment]
It would have been bionic64

**Comment by andyp1per** [comment]
I just found another stack overflow on MacOS using this @peterbarker . It definitely works on mac. Can we just merge on that basis? I don't really care enough about clang on linux to put more effort into it - linux users have valgrind.

**Comment by peterbarker** [comment]
I'd be happy to merge this if non-Mac users never saw it so we didn't get support questions on it.

Failing that, just a "mac users only!" addition to the help text would probably also be sufficient.

**Comment by andyp1per** [comment]
@peterbarker ok added the help comment and change the commit messages and docs

**Comment by peterbarker** [comment]
Thanks @andyp1per - would have been nice to have this on all platforms, but some is better than none.

**Pull request merged**